### ISSUE TITLE
Closing listening socket never terminates

### ISSUE THREAD

**Issue by basvandijk** [yesodweb/wai#691]
At work I noticed that I couldn't terminate my warp-based webservers. Fortunately systemd will kill the process after 1m and 30s however, I don't want to rely on that. I isolated the problem to the following small example program:

https://github.com/basvandijk/warp-close-test

Note how I use `setInstallShutdownHandler` to install a shutdown handler. The shutdown handler will install a signal handler to handle SIGTERM. The handler for SIGTERM will then try to close the listening socket. However this is where the problem is because the close operation never finishes, i.e. I never see "Socket closed." being printed.

It would be great if somebody could reproduce this! When you have Nix installed you can just run `make`.

I expect to see:

```
$ make
...
Sending SIGTERM...
Received SIGTERM. Closing socket...
Socket closed.
```

However I see:
```
$ make
...
Sending SIGTERM...
Received SIGTERM. Closing socket...
Sending SIGKILL
```
* OS name and version

```
$ nixos-version 
18.03pre-git (Impala)
```

```
$ uname -a
Linux bassbox 4.9.101 #1-NixOS SMP Sat May 19 08:27:01 UTC 2018 x86_64 GNU/Linux
```

* The versions of tools you're using (e.g. `stack`, `ghc`).
```
$ cabal --version
cabal-install version 2.0.0.1
compiled using version 2.0.1.0 of the Cabal library 
```

* The versions of dependencies you're using

I tested this from warp-3.0.13.1 to 3.2.22. I also tested this with `network-2.7.0.0`.

```
$ ghc-pkg list
/nix/store/nj3ja270bd5qyvc87kmal9x7w1cjsmns-ghc-8.2.2-with-packages/lib/ghc-8.2.2/package.conf.d
    Cabal-2.0.1.0
    appar-0.1.4
    array-0.5.2.0
    async-2.1.1.1
    attoparsec-0.13.2.2
    auto-update-0.1.4
    base-4.10.1.0
    binary-0.8.5.1
    blaze-builder-0.4.0.2
    bsb-http-chunked-0.0.0.2
    byteorder-1.0.4
    bytestring-0.10.8.2
    bytestring-builder-0.10.8.1.0
    case-insensitive-1.2.0.10
    containers-0.5.10.2
    deepseq-1.4.3.0
    directory-1.3.0.2
    filepath-1.4.1.2
    ghc-8.2.2
    ghc-boot-8.2.2
    ghc-boot-th-8.2.2
    ghc-compact-0.1.0.0
    ghc-prim-0.5.1.1
    ghci-8.2.2
    hashable-1.2.6.1
    haskeline-0.7.4.0
    hoopl-3.10.2.2
    hpc-0.6.0.3
    http-date-0.0.6.1
    http-types-0.9.1
    http2-1.6.3
    integer-gmp-1.0.1.0
    integer-logarithms-1.0.2
    iproute-1.7.2
    network-2.6.3.4
    pretty-1.1.3.3
    primitive-0.6.3.0
    process-1.6.1.0
    psqueues-0.2.5.0
    random-1.1
    rts-1.0
    scientific-0.3.5.2
    semigroups-0.18.4
    simple-sendfile-0.2.27
    stm-2.4.5.0
    streaming-commons-0.1.19
    template-haskell-2.12.0.0
    terminfo-0.4.1.0
    text-1.2.2.2
    time-1.8.0.2
    transformers-0.5.2.0
    unix-2.7.2.2
    unix-compat-0.5.0.1
    unordered-containers-0.2.8.0
    vault-0.3.1.0
    wai-3.2.1.1
    warp-3.2.22
    word8-0.1.3
    xhtml-3000.2.2
    zlib-0.6.1.2
```

**Comment by basvandijk** [comment]
The process is hanging on `close socket` [here](https://github.com/yesodweb/wai/blob/master/warp/Network/Wai/Handler/Warp/Run.hs#L133).

Using `strace` I don't see any `close` syscalls after printing "Received SIGTERM. Closing socket":

```
write(1, "Received SIGTERM. Closing socket"..., 36) = 36
+++ killed by SIGKILL +++
```

so the hang must occur in Haskell. I see that [`close` is using `modifyMVar_`](http://hackage.haskell.org/package/network-2.7.0.0/docs/src/Network.Socket.html#line-1346). Maybe something is taking the `socketStatus` MVar...

**Comment by basvandijk** [comment]
Ok, after debugging the `network` library, I can confirm the program is blocking on the `modifyMVar_` in `close`.

**Comment by kazu-yamamoto** [comment]
> Note how I use `setInstallShutdownHandler` to install a shutdown handler. The shutdown handler will install a signal handler to handle SIGTERM. The handler for SIGTERM will then try to close the listening socket. However this is where the problem is because the close operation never finishes, i.e. I never see "Socket closed." being printed.

`runSettings` ensures that a listening socket will be closed in `bracket`.
Why do you try to close the socket doubly with `setInstallShutdownHandler`?

**Comment by basvandijk** [comment]
Hi Kazu, because in my real application I would like to implement a graceful shut down. So when I receive SIGTERM I would like to:

* stop receiving new connections (closeSocket)
* immediately kill long-polling and websocket connections 
* allow all other established connections to finish (or timeout)

So to implement this I think I need control over when the listening socket is closed. Do you know of other ways to implement the same behaviour?

Also if you're not supposed to call closeSocket why is it provided as an argument to the shutdown handler?

**Comment by kazu-yamamoto** [comment]
So, you should not use `runSettings` but use `runSettingsSocket`. This avoids the conflict. 
You can see how `mighttpd2` implements the graceful shutdown: https://github.com/kazu-yamamoto/mighttpd2/blob/9973d756cb3656fcb37f8a4aae45d5fd3c02d196/src/Server.hs#L119

**Comment by kazu-yamamoto** [comment]
- Warp waits that all connections are closed by default.
- You can set timout for the connections by `setGracefulShutdownTimeout`.
- You should close the listening socket by yourself. You must not use `runSettings` for this purpose.

**Comment by kazu-yamamoto** [comment]
@basvandijk Any feedback?

**Comment by basvandijk** [comment]
As [mentioned in network #326](https://github.com/haskell/network/issues/326#issuecomment-399787104) I just build `warp-close-test` against network-HEAD (I had to adapt `simple-sendfile` and `warp` to the new `network` API which was trivial) and the test finished successfully. I see that in network-HEAD the `withMVar` in `accept` and the `modifyMVar` in `close` are no longer there. That very likely fixed the deadlock.

@kazu-yamamoto I should also probably switch to your recommended way of doing a graceful shutdown but I'm not sure yet how to implement it (probably because I haven't studied `mightttpd2` enough to see how you do it).

I'm not sure what to do with this issue. When network-HEAD is released this will be fixed automatically. Maybe we should leave it open until that happens?

**Comment by kazu-yamamoto** [comment]
As one maintainer of `network`, we will release at least 2.7.x.x to fix the `close` issue ASAP.

**Comment by basvandijk** [comment]
This has been fixed in [network-2.6.3.6](http://hackage.haskell.org/package/network-2.6.3.6).
